### PR TITLE
fix issue#59: add validation check during submission on WEB UI

### PR DESCRIPTION
**Pull request by LeZhang2016**

### CODE REVIEW

File: src/web/web-service/frontend/index.html [hunk 1]
```diff
@@ -163,27 +162,30 @@ <h4 style="margin: 18px 12px 12px 24px">OT Border Router</h4>
               <div layout="row">
                 <md-input-container flex="50">
                   <label>Network Name</label>
-                  <input required name="networkName" ng-model="thread.networkName">
+                  <input required md-maxlength="16" name="networkName" ng-model="thread.networkName">
                   <div ng-messages="threadForm.networkName.$error">
                     <div ng-message="required">This is required.</div>
+                    <div ng-message="md-maxlength">The Netowork Name must be less than 16 characters long.</div>
```
Comment (bukepo) [inline src/web/web-service/frontend/index.html]: `The Netowork Name must be no more than 16 characters!`?

File: src/web/web-service/frontend/index.html [hunk 2]
```diff
@@ -199,27 +201,31 @@ <h4 style="margin: 18px 12px 12px 24px">OT Border Router</h4>
...
+                    <div ng-message-exp="['required', 'min', 'max']">Channel must be in 11~26.</div>
                   </div>
                 </md-input-container>
               </div>
 
               <div layout="row">
                 <md-input-container flex="100">
                   <label>On-Mesh Prefix</label>
-                  <input required name="prefix" ng-model="thread.prefix">
+                  <input required name="prefix" ng-model="thread.prefix" ng-pattern="/^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1})/g">
```
Comment (bukepo) [inline src/web/web-service/frontend/index.html]: Does this regex allows prefix length?
Is link-local address allowed as Thread prefix?

File: src/web/web-service/frontend/index.html (hunk 1 above)
Comment (LeZhang2016) [inline src/web/web-service/frontend/index.html]: yes, in Thread spec v1.1 (page 8-26), there is the interpretation about the network name.

Comment (lanyuwen) [inline src/web/web-service/frontend/index.html]: "less than" or "no more than"?

File: src/web/web-service/frontend/index.html [hunk 3]
```diff
@@ -163,27 +162,30 @@ <h4 style="margin: 18px 12px 12px 24px">OT Border Router</h4>
...
+                    <div ng-message-exp="['maxlength', 'minlength', 'pattern']">The Extended PAN ID must be 16 characters long.</div>
                   </div>
                 </md-input-container>
               </div>
 
               <div layout="row">
                 <md-input-container flex="50">
                   <label>PAN ID</label>
-                  <input required name="panId" ng-model="thread.panId">
+                  <input required name="panId" ng-model="thread.panId" ng-pattern="/^0x[0-9a-fA-F]{4}/g">
```
Comment (bukepo) [inline src/web/web-service/frontend/index.html]: not sure if this regex allows 0x1234567890?

File: src/web/web-service/frontend/index.html [hunk 4]
```diff
@@ -199,27 +201,31 @@ <h4 style="margin: 18px 12px 12px 24px">OT Border Router</h4>
               <div layout="row">
                 <md-input-container flex="50">
                   <label>Network Key</label>
-                  <input required name="networkKey" ng-model="thread.networkKey">
+                  <input required minlength="32" maxlength="32" ng-pattern="/^[0-9a-fA-F]{32}/g" name="networkKey" ng-model="thread.networkKey">
```
Comment (bukepo) [inline src/web/web-service/frontend/index.html]: same as above

File: src/web/web-service/frontend/index.html [hunk 5]
```diff
@@ -199,27 +201,31 @@ <h4 style="margin: 18px 12px 12px 24px">OT Border Router</h4>
...
+                    <div ng-message-exp="['required', 'min', 'max']">Channel must be in 11~26.</div>
                   </div>
                 </md-input-container>
               </div>
 
               <div layout="row">
                 <md-input-container flex="100">
                   <label>On-Mesh Prefix</label>
-                  <input required name="prefix" ng-model="thread.prefix">
+                  <input required name="prefix" ng-model="thread.prefix" ng-pattern="/^((?!(fe80:)|(ff[01][0-9]:))([0-9a-fA-F]{1,4}:)(([0-9a-fA-F]{1,4}:){1,7}):((/[1-9][0-9]?){0,1}))$/">
```
Comment (bukepo) [inline src/web/web-service/frontend/index.html]: does this allows ff05::/64

File: src/web/web-service/frontend/res/js/app.js [hunk 6]
```diff
@@ -212,7 +217,12 @@
         };
 
 
-        $scope.showConfirm = function(ev) {
+        $scope.showConfirm = function(ev, valid) {
+            if (valid === false)
```
Comment (bukepo) [inline src/web/web-service/frontend/res/js/app.js]: if valid is boolean, is it simpler to use !valid here?

File: src/web/web-service/frontend/join.dialog.html [hunk 7]
```diff
@@ -11,18 +11,22 @@
...
+              Network Key must be hexadecimal digits and 32 charactors long.
+            </div>
           </div>
         </md-input-container>
       </div>
       <div layout="row">
         <md-input-container flex="80">
           <label>Prefix</label>
-          <input required name="prefix" ng-model="thread.prefix">
+          <input required name="prefix" ng-model="thread.prefix" ng-pattern="/^((?!(fe80:)|(ff[01][0-9]:))([0-9a-fA-F]{1,4}:)(([0-9a-fA-F]{1,4}:){1,7}):((/[1-9][0-9]?){0,1}))$/">
```
Comment (bukepo) [inline src/web/web-service/frontend/join.dialog.html]: same as above

File: src/web/web-service/frontend/index.html [hunk 8]
```diff
@@ -199,27 +201,31 @@ <h4 style="margin: 18px 12px 12px 24px">OT Border Router</h4>
...
+                    <div ng-message-exp="['required', 'min', 'max']">Channel must be in 11~26.</div>
                   </div>
                 </md-input-container>
               </div>
 
               <div layout="row">
                 <md-input-container flex="100">
                   <label>On-Mesh Prefix</label>
-                  <input required name="prefix" ng-model="thread.prefix">
+                  <input required name="prefix" ng-model="thread.prefix" ng-pattern="/^((?!(fe80:)|(ff[0-9a-fA-F][0-9a-fA-F]:))([0-9a-fA-F]{1,4}:)(([0-9a-fA-F]{1,4}:){1,7}):((/[1-9][0-9]?){0,1}))$/">
```
Comment (bukepo) [inline src/web/web-service/frontend/index.html]: It seems this regex cannot work. Have you tried the following IPs?
```
FF01:0:0:0::/64
FE80:200:5aee::/64
fd11:22::/120
```
I guess we can add regex modifier to perform case-insensitive matching. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions

File: src/web/web-service/frontend/join.dialog.html [hunk 9]
```diff
@@ -11,18 +11,22 @@
...
+              Network Key must be hexadecimal digits and 32 charactors long.
+            </div>
           </div>
         </md-input-container>
       </div>
       <div layout="row">
         <md-input-container flex="80">
           <label>Prefix</label>
-          <input required name="prefix" ng-model="thread.prefix">
+          <input required name="prefix" ng-model="thread.prefix" ng-pattern="/^((?!(fe80:)|(ff[0-9a-fA-F][0-9a-fA-F]:))([0-9a-fA-F]{1,4}:)(([0-9a-fA-F]{1,4}:){1,7}):((/[1-9][0-9]?){0,1}))$/">
```
Comment (bukepo) [inline src/web/web-service/frontend/join.dialog.html]: I think the `/` separator for prefix should be escaped.

File: src/web/web-service/frontend/join.dialog.html [hunk 10]
```diff
@@ -11,18 +11,22 @@
...
           <div ng-messages="threadForm.prefix.$error">
```
Comment (LeZhang2016) [inline src/web/web-service/frontend/join.dialog.html]: the regular expression is composed by three parts:
1. `?!([fF][eE]80:|[fF]{2}[0-9a-fA-F][0-9a-fA-F]:))` filters out link local address and multicast address.
2. `(\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*)` checks valid IPv6 address.
3. `(/(1([01][0-9]|2[0-8]))|/([0-9][0-9]?))? ` checks the prefix length [0, 128].

Comment (bukepo) [inline src/web/web-service/frontend/join.dialog.html]: Should exclude `::` as well.

**Pull request merged**